src/xopen/__init__.py

```diff
@@ -402,6 +402,24 @@ def xopen(filename, mode='r', compresslevel=6, threads=None):
         return _open_xz(filename, mode)
     elif filename.endswith('.gz'):
         return _open_gz(filename, mode, compresslevel, threads)
+    elif mode.startswith("r"):
+        # Test up to the first 6 bytes to detect the file format.
+        with open(filename, "rb") as fh:
+            bs = fh.read(6)
+            if not _PY3:
+                bs = bytearray(bs)
+        if bs[:2] == b'\x1f\x8b':
+            # https://tools.ietf.org/html/rfc1952#page-6
+            return _open_gz(filename, mode, compresslevel, threads)
```

Comment:
Ok, perhaps it’s worth changing this since `flake8` complains abotu `xopen()` becoming too complex. Moving the content-based file type detection into a separate function would take care of this.

Reply:
I ended up refactoring the name detection and content detection into two difference functions. However, `flake8` still thinks the function is too complex. I'm open to other suggestions...